### PR TITLE
Implement `Atomics.pause`

### PR DESCRIPTION
**Pull request by jedel1043**
https://github.com/tc39/proposal-atomics-microwait

### CODE REVIEW

File: core/engine/src/builtins/atomics/mod.rs [hunk 1]
```diff
@@ -512,6 +516,50 @@ impl Atomics {
...
+            // Clamp to u16 so that the main thread cannot block using this.
+            iteration_number as u16
+        };
+
+        // 2. If the execution environment of the ECMAScript implementation supports a signal that the current executing code
+        //    is in a spin-wait loop, send that signal. An ECMAScript implementation may send that signal multiple times,
+        //    determined by iterationNumber when not undefined. The number of times the signal is sent for an integral Number
+        //    N is at most the number of times it is sent for N + 1.
+        for _ in 0..iterations {
+            std::hint::spin_loop();
```
Comment (jasonwilliams) [inline core/engine/src/builtins/atomics/mod.rs:557]: Nice, didn't know about this hint

File: core/engine/src/builtins/atomics/mod.rs [hunk 2]
```diff
@@ -55,8 +55,12 @@ impl IntrinsicObject for Atomics {
             .static_method(Atomics::sub, js_string!("sub"), 3)
             .static_method(Atomics::wait, js_string!("wait"), 4)
             .static_method(Atomics::notify, js_string!("notify"), 3)
-            .static_method(Atomics::bit_xor, js_string!("xor"), 3)
-            .build();
+            .static_method(Atomics::bit_xor, js_string!("xor"), 3);
+
+        #[cfg(feature = "experimental")]
+        let builder = builder.static_method(Atomics::pause, js_string!("pause"), 0);
```
Comment (jasonwilliams) [inline core/engine/src/builtins/atomics/mod.rs:61]: Should the length be 1 if it takes an iteration number?

Comment (jedel1043) [reply core/engine/src/builtins/atomics/mod.rs:61]: It's optional, so the required number of arguments is zero.

File: core/engine/src/builtins/atomics/mod.rs [hunk 3]
```diff
@@ -512,6 +516,50 @@ impl Atomics {
...
+
+            // b. If ℝ(iterationNumber) < 0, throw a RangeError exception.
+            let iteration_number = iteration_number.to_number(context)? as i16;
+            if iteration_number < 0 {
+                return Err(JsNativeError::range()
+                    .with_message("`iterationNumber` must be a positive integer")
+                    .into());
+            }
+
+            // Clamp to u16 so that the main thread cannot block using this.
```
Comment (jasonwilliams) [inline core/engine/src/builtins/atomics/mod.rs:548]: Could you explain what you mean here?

**Review by jasonwilliams — APPROVED** [review]
LGTM, I assume the clamping to u16 is to prevent blocking for too long.

**Comment by jedel1043** [comment]
> LGTM, I assume the clamping to u16 is to prevent blocking for too long.

Yes, blocking for `u32::MAX` is a bit too much (20s), but blocking for `u16::MAX` takes a couple of micros.

**Review by raskad — APPROVED** [review]
Looks good

**Pull request merged**